email_verification/models.py

```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Models for email verification
+"""
+
+# Imports #####################################################################
+
+from datetime import datetime, timedelta, timezone
+from urllib.parse import urljoin
+
+from django.conf import settings
+from django.core.mail import send_mail
+from django.core.urlresolvers import reverse
+from django.db import models
+from django.utils.crypto import get_random_string
+
+
+# Models ######################################################################
+
+class EmailVerification(models.Model):
+    """
+    Verification status for an email address. The main entry point is the
+    `verify` classmethod.
+    """
+    VERIFICATION_CODE_LENGTH = 64
+
+    verification_code_expiry = timedelta(days=7)
+    verification_email_subject = 'Please verify this email address'
+
+    email = models.EmailField(primary_key=True)
+    verified = models.BooleanField(default=False)
+    verification_code = models.CharField(max_length=VERIFICATION_CODE_LENGTH,
+                                         db_index=True)
+    verification_email_date = models.DateTimeField(null=True)
+    verification_date = models.DateTimeField(null=True)
+
+    @classmethod
+    def verify(cls, email, base_url, force_send_email=False):
+        """
+        Verify the given email address. Returns a corresponding
+        `EmailVerification` instance.
+
+        `base_url` should be the domain prefix to include in the verification
+        link, e.g. 'http://example.com/'.
+
+        If an email has already been sent to verify the given address, a new
+        email will not be sent unless `force_send_email` is True.
+        """
+        try:
+            instance = cls.objects.get(email=email)
+        except cls.DoesNotExist:
+            instance = cls(email=email)
+        if force_send_email or not instance.verification_email_sent:
+            instance.send_verification_email(base_url)
+        return instance
+
+    @classmethod
+    def check_verification_code(cls, code):
```

Comment:
@omarkhan I think it would make sense to rename this method to `verify_mail`, to better reflect its main purpose. The fact that it checks the verification code is sort of incidental (it might be doing other checks as well before marking an email address as valid).